src/haddock/modules/topology/topocg/__init__.py

```diff
@@ -1,12 +1,339 @@
-"""Create and manage CNS coarse-grained-atom topology.
+"""Create and manage CNS all-atom topology.
 
-This module is under development...
+The ``[topocg]`` module is dedicated to the generation of CNS compatible
+parameters (.param) and topologies (.psf) for each of the input structures.
+
+It will:
+- Detect missing atoms, including hydrogens
+- Re-build them when missing
+- Build and write out topologies (psf) and coordinates (pdb) files
+
+This module is a pre-requisite to run any downstream modules using CNS.
+Having access to parameters and topology is mandatory for any kind
+of EM/MD related tasks.
+Therefore this is the reason why the module [topocg] is often used as first
+module in a workflow.
+
+Note that for non-standard bio-molecules
+(apart from standard amino-acids, some modified ones, DNA, RNA, ions
+and carbohydrates ... see `detailed list of supported molecules 
+<https://wenmr.science.uu.nl/haddock2.4/library>`_),
+such as small-molecules, parameters and topology must be obtained and provided
+by the user, as there is currently no built-in solution to generate
+them on the fly.
 """
-# to be developed
-# adding dummy interfaces to pass the tests
+
+import operator
+import os
+import re
+from functools import partial
 from pathlib import Path
 
+from haddock.core.typing import FilePath, Optional, ParamDict, ParamMap, Union
+from haddock.libs import libpdb
+from haddock.libs.libcns import (
+    generate_default_header,
+    load_workflow_params,
+    prepare_output,
+    prepare_single_input,
+)
+from haddock.libs.libontology import Format, PDBFile, TopologyFile
+from haddock.libs.libstructure import make_molecules
+from haddock.libs.libontology import PDBFile
+from haddock.libs.libstructure import Molecule
+from haddock.libs.libsubprocess import CNSJob
+from haddock.modules import get_engine
+from haddock.modules.base_cns_module import BaseCNSModule
+
+import haddock.modules.topology.topocg.aa2cg
+
 
 RECIPE_PATH = Path(__file__).resolve().parent
 DEFAULT_CONFIG = Path(RECIPE_PATH, "defaults.yaml")
 
+
+def generate_topology(
+    input_pdb: Path,
+    output_path: str,
+    recipe_str: str,
+    defaults: ParamMap,
+    mol_params: ParamMap,
+    default_params_path: Optional[FilePath] = None,
+    write_to_disk: Optional[bool] = True,
+) -> Union[Path, str]:
+    """Generate a HADDOCK topology file from input_pdb."""
+    # generate params headers
+    general_param = load_workflow_params(**defaults)
+    input_mols_params = load_workflow_params(param_header="", **mol_params)
+    general_param = general_param + input_mols_params
+
+    # generate default headers
+    link, trans_vec, tensor, scatter, axis, water_box = generate_default_header(
+        path=default_params_path
+    )
+
+    # AA to CG
```

Comment:
Does it also generate the required distance restraints for the mapping from CG to AA?

Reply:
Yes, .tbl files are generated but now that i check they are not archived (.tbl.gz)